### DISCUSSION TITLE
Clouditor 2.0 Architecture Discussions

### DISCUSSION THREAD

**Discussion by oxisto** [clouditor/clouditor#206] [General]
# Architecture

- We are aiming for Go as the primary language
- We make use of gRPC to implement the different components / service 
- We expose certain services as a REST API using https://github.com/grpc-ecosystem/grpc-gateway
- We aim to be compliant to the MEDINA architecture with regards to the services
- We want to make a standalone version of Clouditor available, that integrates all components into one binary
- We also want to enable a full blown micro service architecture where the individual components communicate via gRPC and are not necessarily even on the same machine

# Discovery and Assessment

- We want to discover resource in a Cloud-provider agnostic and "semantic" way
- We want to support Azure, AWS and Kubernetes
- We want to offer a way for third party tools to "discover" resources according to our semantic schema. This results into *nodes* in a graph
- We create (MEDINA-compliant) evidences from those graph nodes (in the discovery service) and push them to the assessment service. For example, we can serialise the graph node into JSON and put them into the `resource` property of the evidence. For external tools, we need to specify the format of the evidence `resource` property that the Clouditor assessment service expects (aka our ontology)
- We want to store an up-to-date (as far as possible) snapshot of all Cloud resources that we are assessing. We (currently) are not aiming to store a historical inventory
- We want to design rules that impose constraints on these generic objects, i.e. `ObjectStorage needs to have encryption turned on`
- We make use of the existing Cloud Compliance Language (CCL) to create *assessment results* out of *evidences* according to a *metric* and its *target value(s)*

# Metric Configuration

- TBD: Configuration of target values

cc @garuppel @immqu @LeBogg

**Comment by immqu** [comment]
_We make use of the existing Cloud Compliance Language (CCL) to create assessment results out of evidences according to a metric and its target value(s)_
Update: To create assessment results, we consider using the policy engine OPA with its language Rego instead of our custom CCL; it fits the use case well since authorization decisions are similar to the compliance check we want to do, and it has good overall tool support
